atlasdb-timelock-server/src/test/java/com/palantir/atlasdb/timelock/paxos/PaxosTimestampBoundStoreTest.java

```diff
@@ -0,0 +1,324 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.atlasdb.timelock.paxos;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.mockito.Matchers.anyLong;
+import static org.mockito.Matchers.anyObject;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import javax.annotation.Nullable;
+
+import org.apache.commons.io.FileUtils;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.palantir.common.concurrent.PTExecutors;
+import com.palantir.common.remoting.ServiceNotAvailableException;
+import com.palantir.leader.proxy.ToggleableExceptionProxy;
+import com.palantir.paxos.PaxosAcceptor;
+import com.palantir.paxos.PaxosAcceptorImpl;
+import com.palantir.paxos.PaxosLearner;
+import com.palantir.paxos.PaxosLearnerImpl;
+import com.palantir.paxos.PaxosProposer;
+import com.palantir.paxos.PaxosProposerImpl;
+import com.palantir.paxos.PaxosRoundFailureException;
+import com.palantir.timestamp.MultipleRunningTimestampServiceError;
+
+public class PaxosTimestampBoundStoreTest {
+    private static final int NUM_NODES = 5;
+
+    private static final String LOG_DIR = "testlogs/";
+    private static final String LEARNER_DIR_PREFIX = LOG_DIR + "learner/";
+    private static final String ACCEPTOR_DIR_PREFIX = LOG_DIR + "acceptor/";
+    private static final long TIMESTAMP_1 = 100000;
+    private static final long TIMESTAMP_2 = 200000;
+    private static final long TIMESTAMP_3 = 300000;
+    private static final long FORTY_TWO = 42;
+    private static final PaxosTimestampBoundStore.SequenceAndBound ONE_AND_FORTY_TWO =
+            ImmutableSequenceAndBound.of(1, FORTY_TWO);
+
+    private static final String STILL_RUNNING_MESSAGE =
+            "Some threads are still hanging around! Can't proceed or they might corrupt future tests.";
+    private static final RuntimeException EXCEPTION = new RuntimeException("exception");
+
+    private final ExecutorService executor = PTExecutors.newCachedThreadPool();
+    private final List<PaxosAcceptor> acceptors = Lists.newArrayList();
+    private final List<PaxosLearner> learners = Lists.newArrayList();
+    private final List<AtomicBoolean> failureToggles = Lists.newArrayList();
```

Comment:
Each `ToggleableExceptionProxy` has a reference to an AtomicBoolean, which determines whether to throw an exception or to send the method call through to the real method. See `ToggleableExceptionProxy#invoke()` for how it works.